File: 2677-cousins-in-binary-tree-ii/2677-cousins-in-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def BFS1(root):
            q=deque([root])
            levelsum=[]
            while q:
                l=len(q)
                suml=0
                for _ in range(l):
                    r=q.popleft()
                    suml+=r.val
                    if r.left:
                        q.append(r.left)
                    if r.right:
                        q.append(r.right)
                levelsum.append(suml)
            return levelsum
        def BFS2(root,levelsum):
            q=deque([root])
            root.val=0
            level=1
            while q:
                l=len(q)
                for _ in range(l):
                    r=q.popleft()
                    sibling_sum=0
                    if r.left:
                        sibling_sum+=r.left.val
                    if r.right:
                        sibling_sum+=r.right.val
                    if r.left:
                        r.left.val=levelsum[level]-sibling_sum
                        q.append(r.left)
                    if r.right:
                        r.right.val=levelsum[level]-sibling_sum
                        q.append(r.right)
                level+=1
            return root            
        levelsum=BFS1(root)
        print(levelsum)
        return BFS2(root,levelsum)

File: 2677-cousins-in-binary-tree-ii/2677-cousins-in-binary-tree-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<Integer> BFS1(TreeNode root){
        int i,l,suml,levelsum;
        Queue<TreeNode> q=new LinkedList<>();
        List<Integer> levelSum=new ArrayList<>();
        q.add(root);
        while(!q.isEmpty()){
            l=q.size();
            suml=0;
            for(i=0;i<l;i++){
                TreeNode r=q.poll();
                suml+=r.val;
                if(r.left!=null){
                    q.add(r.left);
                }
                if(r.right!=null){
                    q.add(r.right);
                }
            }
            levelSum.add(suml);
        }
        return levelSum;
    }
    private TreeNode BFS2(TreeNode root,List<Integer> levelSum){
        int i,l,siblingSum,level=1;
        Queue<TreeNode> q=new LinkedList<>();
        root.val=0;
        q.add(root);
        while(!q.isEmpty()){
            l=q.size();
            for(i=0;i<l;i++){
                TreeNode r=q.poll();
                siblingSum=0;
                if(r.left!=null){
                    siblingSum+=r.left.val;
                }
                if(r.right!=null){
                    siblingSum+=r.right.val;
                }
                if(r.left!=null){
                    r.left.val=levelSum.get(level)-siblingSum;
                    q.add(r.left);
                }
                if(r.right!=null){
                    r.right.val=levelSum.get(level)-siblingSum;
                    q.add(r.right);
                }
            }
            level++;
        }
        return root;
    }
    public TreeNode replaceValueInTree(TreeNode root) {
        List<Integer> levelsum=BFS1(root);
        System.out.println(levelsum);
        return BFS2(root,levelsum);
    }
}